### ISSUE TITLE
Link a Projects as a component of another project

### ISSUE THREAD

**Issue by thompson-tomo** [DependencyTrack/dependency-track#3375]
### Current Behavior

User must manually calculate vulnerabilities of all projects being deployed

### Proposed Behavior

I wish for the "add component" button to now have a drop-down of 2 options:

- Component
- Project

When a user selects the project option they are provided UI to select a project and version which already exists in dependency track.

The way my releases work i have multiple micro-services which are bundled together as part of a build pipeline to form 1 installer for the end user. 
With the proposed approach i could select a "project" which reporesents my installed and see all the projects which are contributing to that installer and have install visibility into the health of that release.

### Checklist

- [X] I have read and understand the [contributing guidelines](https://github.com/DependencyTrack/dependency-track/blob/master/CONTRIBUTING.md#filing-issues)
- [X] I have checked the [existing issues](https://github.com/DependencyTrack/dependency-track/issues) for whether this enhancement was already requested

**Comment by valentijnscholten** [comment]
How would this work on SBOM upload for a project that has a dependency on another project? In the SBOM there's must a purl of that other project. But that purl can or cannot be present in DT. And if it is present, maybe the uploader does not want to link them together. Can this feature be implemented without requiring extra information in the SBOM / from the uploader?

**Comment by thompson-tomo** [comment]
I don't think we need to touch/change the SBOM based on the linking being opt-in via clicking the add component button and then selecting the project.

**Comment by valentijnscholten** [comment]
The SBOM is the single source of truth. Any components added later will be removed upon the next SBOM upload. My point was that it should be automateable to be scalable.

**Comment by thompson-tomo** [comment]
Sure I do agree that it would be very nice if the linking could be automated based on purl but establishing that connectivity would be the first key step even if manual.

**Comment by savek-cc** [comment]
Hi!
We're actually also working on a feature like this. It has already been discussed here in other issues (t.b.d. - link them to this one).
Our current working discussion is this:


    DependencyTrack shall be able to use DT projects as components within another project.
    Mapping shall be performed based on CPE and/or PURL assigned to the shared project and listed in the consuming projects SBOM
    DependencyTrack shall offer to optionally resolve components from shared projects. (Setting on consuming project-level)
        "Resolve dependencies of shared projects": fully inherit each sub-component (including vulnerabilities) from the shared project (merge SBOM)
        "Do not resolve dependencies of shared projects": fully inherit every un-audited or affected vulnerability as a vulnerability of the shared project instead of the actual sub-component.
    Option 1)
        Shared project has multiple components with multiple CVE/GHSA/...
        ALL sub-components of the shared components are copied/synced to the consuming project
        Basically ONE combined SBOM - with dependency-tree - with all resolved sub-components is created an analyzed using the standard vulnerability sources.
        Assuming spring-web is a dependency of a shared project:
        Image
    Option 2)
        Shared project has multiple components with multiple CVE/GHSA/...
        List of "open" CVE/GHSA/... are inherited to the consuming project
        Inherited CVE/GHSE/... shall be based on AUDITED vulnerability list of shared project
        All vulnerabilities shall be assigned to the "component" shared project (instead of the original sub-component within the shared project)
        Image

Acceptance Criteria

    Given a "main" project that uses an internal shared component, when that shared component has known vulnerabilities and the user selected to resolve sub-projects, then DepTrack lists all vulnerabilities in the shared component's dependency below the "main" project (current state)
    Given a "main" project that uses an internal shared component, when that shared component has known vulnerabilities and the user selected to not resolve sub-projects, then DepTrack only lists the shared project along with its non-audited/affected vulnerabilities.
    Given a project that uses another project as a component, when a user clicks on the component, then the user is redirected to the project's page (instead of the component page)

Hint: Detect and prevent circular dependencies.

**Comment by thompson-tomo** [comment]
Yes #869 is certainly connected to this, I think we should let this issue focus on making it easy to define that relationship using UI rather than uploading SBOM's.

**Comment by savek-cc** [comment]
DependencyTrack is "API first" - what speaks against identifying a project by assigning it a purl (or cpe if you must) and then including this in an SBOM of a "summary" project?
Using the UI you could just add a component with the same purl as the dt project that you assigned this purl - see https://github.com/DependencyTrack/dependency-track/issues/3375#issuecomment-1891603457 for two possibilities on how to "merge" vulnerability/composition information.

**Comment by thompson-tomo** [comment]
I have no issue with the 2 options mentioned. I am just thinking about how to avoid the need to manually create an sbom for a container project which introduces greater chance of error.

My use case is I perform a release of my application. This app is made of 6 different services with each being a dedicated project in DT. In some releases all the services have a new version but in others only 1 might change hence making it easily definable in the ui is desired. After adding a new version of the master app via UI, the user could simply select the new version of the linked project from a drop-down in the edit component screen.

**Comment by savek-cc** [comment]
I think if you separate UI from actual implementation, you could got at it like this:
1) assign each of your (relevant) project a purl-string (so that is is identifiable)
2) reference those projects from another project
    a) by uploading a bom containing those purls
    b) in the UI have an additional option to reference a project for easier selection; in the end/technically this is the same as adding a component with the matching purl though.

**Comment by thompson-tomo** [comment]
Sure it should certainly appear the same when looking at vulnerabilities & dependency graph. In terms of processing the difference should be that a manually linked project should not be wiped on uploading an sbom.

**Comment by savek-cc** [comment]
I don't think that will work - as you'd need a completely different data model. It's all about SBOMs; If you manually edit that, it will be overwritten with a new upload (as is everything else). And it should - as a new upload/version might also have dependencies to later versions of another "project".

**Comment by thompson-tomo** [comment]
Ok if there is no way to have it so that a manually created component is not wiped on the upload of an sbom could there be a way in which a file (ie sbom) with components is passed to the dotnet tool so that those components can be added?

**Comment by savek-cc** [comment]
I don't know your exact setup - I'd reccomend to take that part of the SBOM (download components as SBOM, just keep the part with the manually added ones) and merge it to the automatically generated one using e.g. CycloneDX-CLI as a workflow.

**Comment by thompson-tomo** [comment]
Yes that would be possible but isn't as convenient as simply passing the path of an existing sbom (downloaded & filtered) to the dotnet tool. Your proposal adds an extra build command & requires an additional tool to be installed in the environment.

**Comment by savek-cc** [comment]
Your suggestion isn't usable in an automated way (via the API) at all though - and updating reference projects versions requires a lot of manual work.

**Comment by thompson-tomo** [comment]
My suggestion is the same as yours except I am proposing the dotnet cli tool does the merging rather introduce an additional tool

**Comment by davidecavestro** [comment]
> identifying a project by assigning it a purl (or cpe if you must) and then including this in an SBOM of a "summary" project

I'd like to do this, but I'm not sure if it works with the latest snapshots or if there's a feature missing instead.

**Comment by robertlagrant** [comment]
The latest comments seem to be debating how to incorporate an existing component's SBOM into a new component's SBOM. Would it be possible to pass a pURL (say) in a SBOM and have DT match it with the pURL of an existing uploaded component at runtime? So the "merge" happens in DT?

**Comment by m4nch0t** [comment]
Hello,

I'm very interested in this feature and have been following the discussion. I'd like to share my perspective and a potential approach, though I may have misunderstood some of the previous proposals.

My concern is with how we handle shared components across multiple projects. Let's consider an example where we use the same **Prometheus** image in two different projects: **Project A** and **Project B**.

To track vulnerabilities for Prometheus, we might create a dedicated "Prometheus" project in Dependency-Track.

#### Current Approaches and Challenges

1.  **Merging SBOMs**: If we merge the SBOM of `Project A` with the SBOM of the `Prometheus` project, we lose the global visibility of vulnerabilities in the main dashboard. It becomes difficult to distinguish which vulnerabilities originate from which sub-project or component.

2.  **Duplicating Projects**: To maintain this visibility, we could duplicate the Prometheus project for each use case (e.g., `Prometheus-for-ProjectA`, `Prometheus-for-ProjectB`). However, this creates redundancy and a poor user experience, as we would have multiple projects for the exact same component.

3.  **Hierarchical SBOMs**: To be perfectly accurate, we would need to modify the SBOMs to reflect a hierarchical dependency graph, where the project name is not the same as `metadata/component/name`. This seems overly complex for what we are trying to achieve. (The dependency graph has xxx entries, but the project (metadata.component node of the BOM) is not one of them; Graph will be incomplete because it is not possible to determine its root)

#### Proposed Solution: Project Aliasing or Symbolic Linking

A solution might be to introduce a concept of "aliasing" or "symbolic linking" between projects. And like collection logic we could have a merge logic (with or without hierarchical merge).  

With this approach, we could have a single, canonical project for **Prometheus:v1**. Then, within **Project A** and **Project B**, we could create a "link" or "alias" to this canonical project.

This would allow us to:

  * **Maintain a single source of truth** for the vulnerabilities of `Prometheus:v1`.
  * **Preserve the project hierarchy** and visibility in the main dashboard.
  * **Avoid data duplication** and project clutter.

Here is a simple diagram to illustrate this concept:

```
+----------------+      +----------------+      +-----------------------+
|    Project A   |----->| Prometheus:v1  |<-----|      Project B        |
| (alias)        |      | (canonical)    |      | (alias)               |
+----------------+      +----------------+      +-----------------------+
        |                      ^                      |
        |                      |                      |
        +----------------------+----------------------+
                               |
                        Vulnerability Analysis
```

This way, the vulnerability analysis is performed once on the canonical project, and the results are reflected in all projects that link to it.

Apologies for the noise if this suggestion is out of scope for the current discussion. <3

**Comment by m4nch0t** [comment]
Hello,
Any updates on this one?

**Comment by davidecavestro** [comment]
#5845 slightly overlaps this issue
